### PR TITLE
stop using multiple YAML documents

### LINKED ISSUE #88: don't use multiple yaml documents
Currently, multiple protocols can be specified by putting them into multiple documents in the same yaml file:

``` yaml
protocol:
  - foo
  - bar
---
protocol:
  - bar
  - baz
```

But for newcomers this seems to be confusing and is maybe generally more an obscure yaml feature? So maybe it'd be better to just switch to:

``` yaml
- protocol:
  - foo
  - bar
- protocol:
  - bar
  - baz
```

The question is whether we then want to disallow specifying a single protocol as a top-level array of steps like this:

``` yaml
- foo
- bar
```

Because that'd be hard to parse and confusing.

cc: @xasopheno

### PR DESCRIPTION
**Pull request by soenkehahn**
Fixes #88.

### CODE REVIEW

**Comment by soenkehahn** [comment]
After this PR users can specify protocols like this:

``` yaml
# a single protocol
protocol:
  - step1
  - step2
```

``` yaml
# multiple protocols as an array
- protocol:
  - step1
  - step2
- protocol:
  - step1
  - step2
```

``` yaml
# multiple protocols in an object in a `protocols` field
protocols:
  - protocol:
    - step1
    - step2
  - protocol:
    - step1
    - step2
```

This last option makes new top-level fields possible, e.g.:

``` yaml
interpreter: python3
protocols:
  - protocol:
    - step1
    - step2
  - protocol:
    - step1
    - step2
```

See #89 and #72.

Should we change the names for `protocols` and/or `protocol`? For example like this?

``` yaml
interpreter: python3
tests:
  - steps:
    - step1
    - step2
  - steps:
    - step1
    - step2
```

cc: @matthandlersux @xasopheno

**Comment by matthandlersux** [comment]
WRT changing names for `protocols`, i think there will be a natural and obvious update to the keywords that will be more obvious when we get further along (eg. when building a protocol for a larger/more complicated deploy scripts). so i don't have much of an opinion for changing it right now.

File: src/protocol/mod.rs [hunk 1]
```diff
@@ -218,16 +218,36 @@ impl Protocol {
...
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Protocols {
+    pub protocols: Vec<Protocol>,
+}
 
-    fn parse(yaml: Yaml) -> R<Protocol> {
+impl Protocols {
+    fn from_array(array: &[Yaml]) -> R<Protocols> {
```
Comment (data-pup) [inline src/protocol/mod.rs]: I might call this `try_from_array`, but that's just crustacean convention

File: src/protocol/mod.rs [hunk 2]
```diff
@@ -383,6 +420,67 @@ mod load {
         );
         Ok(())
     }
+
+    #[test]
+    fn allows_to_specify_multiple_protocols() -> R<()> {
```
Comment (data-pup) [inline src/protocol/mod.rs]: nice tests!

File: README.md [hunk 3]
```diff
@@ -83,28 +84,20 @@ protocol:
...
-Similarly, you can specify the whole protocol as a list of steps. So the above protocol can also be written as this:
-
-``` yaml
-- git add .
-- git push
-```
-
 #### Multiple protocols
 
-YAML supports multiple documents in one file. This can be used to specify multiple protocols:
```
[the remark this answers is not in the document]
Comment (soenkehahn) [reply README.md]: Done.

File: check-protocols-in-docker.sh.protocols.yaml [hunk 4]
```diff
@@ -1,17 +1,16 @@
...
+- arguments: /test/script
+  protocol:
+    - command: /usr/bin/basename /test/script
+      stdout: script
+    - command: /usr/bin/basename /test/script
+      stdout: script
+    - command: /usr/bin/basename /test/script
+      stdout: script
+    - /usr/bin/docker run --rm -it
+      --cap-add=SYS_PTRACE
```
[the remark this answers is not in the document]
Comment (soenkehahn) [reply check-protocols-in-docker.sh.protocols.yaml]: Maybe, but that should be in a different PR, since argument parsing didn't change in this one. #NoScopeCreep

File: src/protocol/mod.rs [hunk 5]
```diff
@@ -237,44 +257,59 @@ impl Protocol {
...
         Ok(result.into_iter().next().unwrap())
     }
 
     #[test]
     fn reads_a_protocol_from_a_sibling_yaml_file() -> R<()> {
         assert_eq!(
-            test_parse(
+            test_parse_one(
                 r##"
-                    |- /bin/true
```
[the remark this answers is not in the document]
Comment (soenkehahn) [reply src/protocol/mod.rs]: Yeah, good idea. Done.
